File: lintcode/959.k-spaced-array-sorting.py
class Solution:
    """
    @param arr: The K spaced array
    @param k: The param k
    @return: Return the sorted array
    """

    def getSortedArray(self, arr, k):
        res = []
        pq = []

        import heapq

        upper = min(len(arr), k)
        for i in range(0, upper):
            heapq.heappush(pq, (arr[i], i))

        while len(pq) > 0:
            v, idx = heapq.heappop(pq)

            res.append(v)
            new_idx = idx + k
            if new_idx < len(arr):
                heapq.heappush(pq, (arr[new_idx], new_idx))

        return res


if __name__ == '__main__':
    Solution().getSortedArray([4, 0, 5, 3, 10], 2)

File: lintcode/959.k-spaced-array-sorting.cpp
#include <algorithm>
#include <queue>
#include <utility>
#include <vector>
using namespace std;
class Solution {
public:
  /**
   * @param arr: The K spaced array
   * @param k: The param k
   * @return: Return the sorted array
   */
  vector<int> getSortedArray(vector<int> &arr, int k) {
    priority_queue<pair<int, int>, vector<pair<int, int>>,
                   greater<pair<int, int>>>
        q;
    for (int i = 0; i < arr.size() && i < k; ++i) {
      q.push({arr[i], i});
    }

    vector<int> res;
    while (!q.empty()) {
      auto p = q.top();
      q.pop();
      res.push_back(p.first);
      if (p.second + k < arr.size()) {
        q.push({arr[p.second + k], p.second + k});
      }
    }
    return res;
  }
};
